src/dmhandler.ts

```diff
@@ -0,0 +1,398 @@
+import {DiscordBridgeConfig} from "./config";
+import {DiscordClientFactory} from "./clientfactory";
+import {
+    Bridge,
+} from "matrix-appservice-bridge";
+import {DiscordStore} from "./store";
+import {DbDmRoom} from "./db/dbdatadmroom";
+import { Message, Client, User, DMChannel, Channel, GroupDMChannel, Util } from "discord.js";
+import { DMRoom } from "./dmroom";
+import { MessageProcessor } from "./messageprocessor";
+import { MatrixEventProcessor } from "./matrixeventprocessor";
+import { Log } from "./log";
+import { UserSyncroniser } from "./usersyncroniser";
+
+const log = new Log("DMHandler");
+
+const NOT_ENABLED_ERROR =
+`
+The Discord bridge is not enabled for DM support, so you cannot message this Discord user.
+You may leave this room.
+`;
+
+const NOT_DM_ROOM =
+`
+Inviting Discord user's to rooms is not supported (this room contains more than two members).
+`;
+
+const NOT_PUPPETED =
+`
+Your account is not puppeted, so you cannot talk to Discord user's privately.
+You may leave this room.
+`;
+
+const MAX_MEMBERS_FOR_DM = 9;
+
+interface InviteResult {
+    valid: boolean;
+    message: string;
+}
+/**
+ * Handler class that forwards memberstate and messages between Matrix and Discord DMs.
+ * This class does not handle direct messaging rules or message formatting which
+ * should be handled in DMRoom.
+ */
+export class DMHandler {
+    private dmRooms: DMRoom[];
+    private discordToUserIdMap: Map<string, string>;
+    private messageProcessor: MessageProcessor;
+    private matrixEventProcessor: MatrixEventProcessor;
+    constructor (
+        private config: DiscordBridgeConfig,
+        private bridge: Bridge,
+        private clientFactory: DiscordClientFactory,
+        private store: DiscordStore,
+        private userSync: UserSyncroniser,
+        ) {
+        this.dmRooms = [];
+        this.messageProcessor = new MessageProcessor({
+            domain: bridge.opts.domain,
+            bot: null, // No emoji support.
```

Comment:
discord nitro people can use custom emotes in DMs - they should relay correctly, too! Perhaps just re-use the other message processor?